File: frontend/src/types/task.ts
import type { DependencyType, TaskStatus } from "./common";
import type { ChecklistItem } from "./checklist";
import type { TaskDependency } from "./dependency";
import type { User } from "./user";

/** Задача проекта — основной элемент реестра задач и диаграммы Ганта. */
export interface Task {
  /** Уникальный идентификатор задачи. */
  id: string;
  /** Идентификатор проекта, которому принадлежит задача. */
  projectId: string;
  /** Идентификатор спринта, к которому привязана задача (может отсутствовать). */
  sprintId: string | null;
  /** Идентификатор родительской задачи для построения иерархии WBS (подзадачи). */
  parentTaskId: string | null;
  /** Контрольная точка, к которой ведёт задача; null — задача вне вех. */
  milestoneId: string | null;
  /** Короткий читаемый код задачи. */
  code: string;
  /** Номер задачи в иерархической структуре работ (WBS). */
  wbsNumber: string;
  /** Название задачи. */
  title: string;
  /**
   * Ответственный исполнитель задачи.
   * Бэкенд присылает `null`, если исполнитель не назначен (см. `newUserDTOPtr`
   * в `backend/internal/api/dto.go`), а не опускает поле.
   */
  assignee: User | null;
  /** Текущий статус задачи. */
  status: TaskStatus;
  /** Признак того, что задача является вехой (точкой с нулевой длительностью) на диаграмме Ганта. */
  isMilestone: boolean;
  /** Плановая дата начала задачи. */
  startDate: string;
  /** Плановая дата окончания задачи. */
  endDate: string;
  /** Продолжительность задачи в календарных днях. */
  durationCalendarDays: number;
  /** Продолжительность задачи в рабочих днях согласно календарю проекта. */
  durationWorkingDays: number;
  /** Процент фактического выполнения задачи. */
  progressPercent: number;
  /** Вес задачи в общем прогрессе проекта, %. */
  weightPercent: number;
  /** Признак того, что задача лежит на критическом пути (нулевой резерв времени). */
  isCriticalPath: boolean;
  /** Резерв (буфер) времени задачи в днях, доступный для компенсации сдвигов. */
  bufferDays: number;
  /** Отклонение факта от плана в днях. Положительное — опережение, отрицательное — отставание. */
  planVsActualDeviationDays: number;
}

/** Связь задачи с внешней системой отслеживания (например, GitLab issue). */
export interface TaskExternalLink {
  /** Название внешней системы. */
  provider: string;
  /** Ссылка на объект во внешней системе. */
  url: string;
  /** Идентификатор объекта во внешней системе. */
  referenceId: string;
  /** Признак того, что данные синхронизированы с внешней системой. */
  synced: boolean;
}

/** Полная карточка задачи со связями, чек-листом DoD и метаданными для экрана редактирования. */
export interface TaskDetail extends Task {
  /** Развёрнутое описание задачи. */
  description?: string;
  /** Связь задачи с внешней системой отслеживания (например, GitLab issue). */
  externalLink: TaskExternalLink | null;
  /** Критерии приёмки (Definition of Done) задачи. */
  checklist: ChecklistItem[];
  /** Задачи-предшественники, от которых зависит текущая задача. */
  predecessors: TaskDependency[];
  /** Задачи-последователи, зависящие от текущей задачи. */
  successors: TaskDependency[];
  /** Количество комментариев к задаче. */
  commentsCount: number;
}

/** Связь с задачей-предшественником, создаваемая одновременно с задачей. */
export interface TaskPredecessorInput {
  /** Идентификатор задачи-предшественника. */
  taskId: string;
  /** Тип связи между задачами. */
  type: DependencyType;
  /** Задержка (лаг) в днях между задачами. */
  lagDays?: number;
}

/** Данные для создания новой задачи. */
export interface TaskCreateRequest {
  /** Идентификатор спринта, к которому привязывается задача. */
  sprintId?: string;
  /** Идентификатор родительской задачи (для создания подзадачи в WBS). */
  parentTaskId?: string;
  /** Контрольная точка, к которой ведёт задача. */
  milestoneId?: string;
  /** Название задачи. */
  title: string;
  /** Описание задачи. */
  description?: string;
  /** Идентификатор назначаемого ответственного. */
  assigneeId?: string;
  /** Начальный статус задачи. */
  status?: TaskStatus;
  /** Признак того, что создаваемая задача является вехой. */
  isMilestone?: boolean;
  /** Плановая дата начала задачи. */
  startDate: string;
  /** Плановая дата окончания задачи. */
  endDate: string;
  /** Вес задачи в общем прогрессе проекта, %. */
  weightPercent?: number;
  /** Список связей с задачами-предшественниками, создаваемых одновременно с задачей. */
  predecessors?: TaskPredecessorInput[];
}

/** Набор полей задачи, доступных для частичного обновления (PATCH). */
export interface TaskUpdateRequest {
  /** Новый спринт задачи. */
  sprintId?: string;
  /** Новое название задачи. */
  title?: string;
  /** Новое описание задачи. */
  description?: string;
  /**
   * Идентификатор нового ответственного.
   * `null` снимает ответственного, отсутствие поля оставляет прежнего —
   * на бэкенде это `*uuid.UUID` (см. `taskUpdateRequest` в `tasks.go`).
   */
  assigneeId?: string | null;
  /** Новый статус задачи. */
  status?: TaskStatus;
  /** Новая дата начала задачи. */
  startDate?: string;
  /** Новая дата окончания задачи. */
  endDate?: string;
  /** Новый процент выполнения задачи. */
  progressPercent?: number;
  /** Новый вес задачи в общем прогрессе проекта, %. */
  weightPercent?: number;
  /**
   * Контрольная точка задачи.
   * Поле не передано — не менять, `null` — отвязать от вехи
   * (на бэкенде это различают `MilestoneID` + `MilestoneIDSet`).
   */
  milestoneId?: string | null;
}

/** Параметры фильтрации списка задач проекта. */
export interface TaskListParams {
  /** Ограничить выборку задачами конкретного спринта. */
  sprintId?: string;
  /** Ограничить выборку задачами конкретного ответственного. */
  assigneeId?: string;
  /** Фильтр по статусу задачи. */
  status?: TaskStatus;
  /** Если true — вернуть только задачи, лежащие на критическом пути. */
  criticalPathOnly?: boolean;
  /** Только задачи с отклонением от плана (в зоне риска). */
  risksOnly?: boolean;
  /** Если true — вернуть только задачи текущего пользователя. */
  myTasksOnly?: boolean;
  /** Подстрока для поиска по названию или коду задачи. */
  search?: string;
}

File: backend/internal/models/task.go
package models

import "github.com/google/uuid"

// Task — задача проекта, основной элемент реестра и диаграммы Ганта.
//
// В таблице лежит только то, что задал пользователь или пересчитал сервис сдвига.
// Производные величины — длительность, признак критического пути, отклонение
// план/факт — считаются на лету в internal/schedule: хранить их значило бы
// поддерживать их в согласованном виде после каждой правки любой связанной задачи.
type Task struct {
	Model

	ProjectID uuid.UUID `gorm:"type:uuid;not null;index"`
	Project   *Project  `gorm:"foreignKey:ProjectID;constraint:OnDelete:CASCADE"`

	// SprintID — привязка к спринту, необязательная.
	SprintID *uuid.UUID `gorm:"type:uuid;index"`
	Sprint   *Sprint    `gorm:"foreignKey:SprintID;constraint:OnDelete:SET NULL"`

	// ParentTaskID — родитель в иерархии работ (WBS). Ссылка на ту же таблицу:
	// удаление родительской задачи уносит её подзадачи.
	ParentTaskID *uuid.UUID `gorm:"type:uuid;index"`
	ParentTask   *Task      `gorm:"foreignKey:ParentTaskID;constraint:OnDelete:CASCADE"`

	// Code — читаемый код задачи («TASK-004»), уникален в пределах проекта.
	Code string `gorm:"type:varchar(32);not null;index"`
	// WBSNumber — номер в иерархической структуре работ («4.2»).
	WBSNumber string `gorm:"type:varchar(32)"`
	// OrderIndex задаёт стабильный порядок задач в реестре и на диаграмме:
	// без него postgres возвращает строки в произвольном порядке и Гант «прыгает».
	OrderIndex int `gorm:"not null;default:0"`

	Title       string `gorm:"type:varchar(255);not null"`
	Description string `gorm:"type:text"`

	AssigneeID *uuid.UUID `gorm:"type:uuid;index"`
	Assignee   *User      `gorm:"foreignKey:AssigneeID;constraint:OnDelete:SET NULL"`

	Status TaskStatus `gorm:"type:varchar(20);not null;index"`
	// IsMilestone — веха на диаграмме: точка нулевой длительности, StartDate == EndDate.
	IsMilestone bool `gorm:"not null;default:false"`

	StartDate Date `gorm:"type:date;not null"`
	EndDate   Date `gorm:"type:date;not null"`

	// Фактические даты заполняются при переходе задачи в работу и при закрытии.
	// Из них считается planVsActualDeviationDays для режима план/факт.
	ActualStartDate *Date `gorm:"type:date"`
	ActualEndDate   *Date `gorm:"type:date"`

	ProgressPercent int     `gorm:"not null;default:0"`
	WeightPercent   float64 `gorm:"not null;default:0"`

	// Связь с внешним трекером хранится плоскими колонками: вложенный объект
	// в ответе собирает DTO, а отдельная таблица ради четырёх полей избыточна.
	ExternalProvider    string `gorm:"type:varchar(50)"`
	ExternalURL         string `gorm:"type:text"`
	ExternalReferenceID string `gorm:"type:varchar(64)"`
	ExternalSynced      bool   `gorm:"not null;default:false"`

	// Ниже — производные поля из комментария выше. gorm:"-" исключает их из
	// миграции и запросов: сервис задач (internal/service, CPM-движок в
	// internal/schedule) заполняет их при чтении, а не хранит.
	//
	// BufferDays раньше был обычной хранимой колонкой со значением по
	// умолчанию 0; она осталась в БД неиспользуемой (AutoMigrate не удаляет
	// колонки) — это ожидаемо и безвредно, не полагаться на её содержимое.
	DurationCalendarDays int  `gorm:"-"`
	DurationWorkingDays  int  `gorm:"-"`
	IsCriticalPath       bool `gorm:"-"`
	BufferDays           int  `gorm:"-"`
}

// TableName фиксирует имя таблицы.
func (Task) TableName() string { return "tasks" }

// PlanVsActualDeviation — отклонение факта от плана в днях; положительное
// значение — опережение графика, отрицательное — отставание.
//
// Полноценный расчёт (earned value) требует CPM-движка, которого пока нет,
// поэтому для задач без фактических дат используется упрощённая эвристика:
// просрочка планового конца без ActualEndDate уже читается как отставание.
func (t Task) PlanVsActualDeviation() int {
	if t.ActualEndDate != nil {
		return t.ActualEndDate.DaysUntil(t.EndDate)
	}
	if today := Today(); today.After(t.EndDate) {
		return -t.EndDate.DaysUntil(today)
	}
	return 0
}

// TaskDependency — связь между двумя задачами по методу критического пути.
//
// Направление задаётся парой полей, а не флагом: предшественник и последователь
// не взаимозаменяемы, и для типов SS/FF/SF порядок определяет смысл связи.
type TaskDependency struct {
	Model

	// Пара уникальна: дублирующие связи между теми же задачами запрещены.
	// Запрет самозависимости и циклов проверяет сервис — выразить их
	// ограничением таблицы нельзя.
	PredecessorTaskID uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_task_dependencies_pair"`
	SuccessorTaskID   uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_task_dependencies_pair"`

	PredecessorTask *Task `gorm:"foreignKey:PredecessorTaskID;constraint:OnDelete:CASCADE"`
	SuccessorTask   *Task `gorm:"foreignKey:SuccessorTaskID;constraint:OnDelete:CASCADE"`

	Type DependencyType `gorm:"type:varchar(2);not null"`
	// LagDays — задержка между задачами; отрицательное значение означает опережение.
	LagDays int `gorm:"not null;default:0"`
}

// TableName фиксирует имя таблицы.
func (TaskDependency) TableName() string { return "task_dependencies" }
